### ISSUE TITLE
Code previews in diagnostics

### ISSUE THREAD

**Issue by ghost** [dotnet/fsharp#14832]
**Is your feature request related to a problem? Please describe.**

F#'s diagnostics are currently single-line brief descriptions with a span and a file id.

![image](https://user-images.githubusercontent.com/93856041/222218380-4933c9d9-6ec5-4837-9751-9a36984c2f1b.png)

This is quite tedious, especially if you're actively working with the dotnet CLI through a terminal. It doesn't really give you an insight on *where* the issue occurs and forces the developer into spawning an editor and finding the location manually.

**Describe the solution you'd like**

The output containing a preview into the source code, with additional metadata and information pointed across spans. Even in simple cases, this is already quite an improvement. But in more complex cases, it can be particularly helpful. 

Consider a hypothetical diagnostic such as this (Image : zesterer/ariadne, diagnostic library in Rust)
![image](https://user-images.githubusercontent.com/93856041/222220425-2c84a282-1782-4c1e-8934-8f2e52e4f897.png)

But this also inherently supports and encourages better, more supporting diagnostics into the compiler.

(Image : Rustc suggesting the proper keyword to use)

![image](https://user-images.githubusercontent.com/93856041/222220124-af3b5777-49ba-4f48-b881-c454561d2d29.png)

**Additional Context**

I would be most glad to implement this into the F# compiler. Bringing my (relatively small, but not insignificant) experience working around with Rustc's diagnostics and working with Ariadne into the F# compiler. =)

**Comment by baronfel** [comment]
The larger problem here is that MSBuild is the one reporting and translating the diagnostics, and there's no facility in MSBuild for 'nice' diagnostics. The MSBuild team (paging @rainersigwald for interest) would love to enable this kind of thing, but there's a multi step process to get there:

* in the MSBuild API for diagnostics expose a way of providing structured data from a diagnostic report (e.g. span ranges that you'd like to highlight for arrows, directives for placeholders in diagnostic text that might want to be highlighted, etc)
* in the MSBuild console logger (we're working on a new, VT-aware one called LiveLogger right now) understand these enhanced diagnostics and print them out in a reasonable way.

separately we have to consider the impact on API consumers of MSBuild diagnostics like VS and VS Code.

**Comment by gimbles** [comment]
> separately we have to consider the impact on API consumers of MSBuild diagnostics like VS and VS Code.

I'm not aware of how VS/VSCode currently communicate, but generally the compiler would emit some form of (non-textual, structural) representation and the tooling would consume that?

**Comment by baronfel** [comment]
yes - you're right that the compilers need to supply all the relevant information as the 'baseline' layer. But from the MSBuild integration perspective, the F# compiler shells out to the CLI version of the compiler, which emits errors in a format that MSBuild can then parse into MSBuild's internal representation of a diagnostic, which MSBuild then emits to the user after yet another transformation. You can see this in the [`Fsc` task](https://github.com/dotnet/fsharp/blob/main/src/FSharp.Build/Fsc.fs) in this repo.  So that's not ideal from a presentation perspective.
